oop-3-inheritance_nd_visibility/index.md

```diff
@@ -0,0 +1,454 @@
+            this.price = price;
+    }
+
+
+    public String getName() {
+        return name;
+    }
+
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+
+    public String getCategory() {
+        return category;
+    }
+
+
+    public void setCategory(String category) {
+        this.category = category;
+    }
+
+
+    @Override
+    public String toString() {
+        return "Product{" +
+               "price=" + price +
+               ", name='" + name + '\'' +
+               ", category='" + category + '\'' +
+               '}';
+    }
+
+}
+```
+
+### What if I have many laptops, and I need to make sure that none of them has its price to be less than 100?
+
+#### Old solution (bad)
+
+```java
+package com.redi.j2;
+
+public class Laptop {
+
+    private float price;
+
+    private String name;
+
+    private String category = "Electrionics";
+
+
+    public float getPrice() {
+        return price;
+    }
+
+
+    protected void setPrice(float price) {
+        if (price < 100)
+            System.out.println("I can't be that cheap");
+        else
+            this.price = price;
+    }
+
+
+    public String getName() {
+        return name;
+    }
+
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+
+    public String getCategory() {
+        return category;
+    }
+
+
+    public void setCategory(String category) {
+        this.category = category;
+    }
+
+
+    @Override
+    public String toString() {
+        return "Product{" +
+               "price=" + price +
+               ", name='" + name + '\'' +
+               ", category='" + category + '\'' +
+               '}' + "I'm also a a laptop btw!";
+    }
+
+```
+
+### new solution (nice!)
+
+Why don't we use inheritance?
+
+```java
+package com.redi.j2;
+
+public class laptop extends Product {
+
+    public laptop() {
+        setCategory("electronics");
+    }
+
+
+    @Override
+    public void setPrice(float price) {
+        if (price < 100) {
+            System.out.println("I can't be THAT cheap!!");
+        }
+        else {
+            super.setPrice(price);
+        }
+    }
+
+
+    @Override
+    public String toString() {
+        return super.toString() + "I am also a laptop btw!";
+    }
+}
+```
+
+### Think of inheritance as an "is-a" relationship
+
+In Java, inheritance is an is-a relationship. That is, we use inheritance only if there exists an is-a relationship between two classes. For example,
+
+Laptop is a Prouct Car is a Vehicle Orange is a Fruit
+
+## Why do we need it?
+
+- The most important use of inheritance in Java is code reusability. The code that is present in the parent class can be directly used by the child
+  class.
+- Method overriding is also known as runtime polymorphism. Hence, we can achieve Polymorphism in Java with the help of inheritance.
+
+## Have we seen it before?
+
+**YES!!** Since day one actually. Remember this example from the very first lesson?
```

Comment:
Maybe it's just me, but I failed to understand the point. How does the example relates to overriding (or inheritance) ?

Reply:
I see now!

Reply:
Maybe the `ArrayList` would be a good example of inheritance?
```java
public class ArrayList<E> extends AbstractList<E>
        implements List<E>, RandomAccess, Cloneable, java.io.Serializable
```

Reply:
Maybe in class I can go through all other data structures and have a very brief look over their documentation.